src/Sarif.Multitool/Rules/SARIF2017.LocationsMustHaveRequiredProperties.cs

```diff
@@ -0,0 +1,99 @@
+using System.Diagnostics;
+using System.Linq;
+
+using Microsoft.Json.Pointer;
+
+namespace Microsoft.CodeAnalysis.Sarif.Multitool.Rules
+{
+    public class LocationsMustHaveRequiredProperties : SarifValidationSkimmerBase
+    {
+        /// <summary>
+        /// SARIF2017
+        /// </summary>
+        public override string Id => RuleId.LocationsMustHaveRequiredProperties;
+
+        /// <summary>
+        /// Each result location must provide the property 'physicalLocation.artifactLocation.uri'.
+        /// The GitHub Developer Security Portal will not display a result whose location does not
+        /// provide the URI of the artifact that was analyzed.
+        /// </summary>
+        public override MultiformatMessageString FullDescription => new MultiformatMessageString { Text = RuleResources.SARIF2017_LocationsMustHaveRequiredProperties_FullDescription_Text };
+
+        protected override IEnumerable<string> MessageResourceNames => new string[] {
+            nameof(RuleResources.SARIF2017_LocationsMustHaveRequired_Properties_Error_NoLocationsArray_Text),
+            nameof(RuleResources.SARIF2017_LocationsMustHaveRequired_Properties_Error_EmptyLocationsArray_Text),
+            nameof(RuleResources.SARIF2017_LocationsMustHaveRequired_Properties_Error_Default_Text)
+        };
+
+        public override FailureLevel DefaultLevel => FailureLevel.Error;
+
+        public override bool EnabledByDefault => false;
+
+        protected override void Analyze(Result result, string resultPointer)
+        {
+            if (result.Locations == null)
+            {
+                // {0}: The 'locations' property is absent. The GitHub Developer Security Portal
+                // will not display a result unless it provides a location that specifies the URI
+                // of the artifact that contains the result.
+                LogResult(
+                    resultPointer,
+                    nameof(RuleResources.SARIF2017_LocationsMustHaveRequired_Properties_Error_NoLocationsArray_Text),
+                    SarifPropertyName.Locations);
+                return;
+            }
+
+            string locationsPointer = resultPointer.AtProperty(SarifPropertyName.Locations);
+            if (!result.Locations.Any())
+            {
+                // {0}: The 'locations' array is empty. The GitHub Developer Security Portal will
+                // not display a result unless it provides a location that specifies the URI of the
+                // artifact that contains the result.
+                LogResult(
+                    locationsPointer,
+                    nameof(RuleResources.SARIF2017_LocationsMustHaveRequired_Properties_Error_EmptyLocationsArray_Text));
+                return;
+            }
+
+            for (int i = 0; i < result.Locations.Count; i++)
+            {
+                Location location = result.Locations[i];
+                string missingProperty = null;
+                string jsonPointer = null;
```

Comment:
>string jsonPointer = null [](start = 16, length = 25)

I would start jsonPointer as LocationPointer.AtIndex(i), since every place you have that. Then, update the other places below to specific cases when necessary.

Reply:
I chose to do it this way because otherwise, we unnecessarily calculate `.AtIndex(i)` for every location in every result in the log file. This way, we only calculate it in the rare case where the rule fires.

---
In reply to: [463585630](https://github.com/microsoft/sarif-sdk/pull/2021#discussion_r463585630) [](ancestors = 463585630)